obstore/python/obstore/fsspec.py

```diff
@@ -0,0 +1,159 @@
+"""Fsspec integration.
+
+The underlying `object_store` Rust crate cautions against relying too strongly on stateful filesystem representations of object stores:
+
+> The ObjectStore interface is designed to mirror the APIs of object stores and not filesystems, and thus has stateless APIs instead of cursor based interfaces such as Read or Seek available in filesystems.
+>
+> This design provides the following advantages:
+>
+> - All operations are atomic, and readers cannot observe partial and/or failed writes
+> - Methods map directly to object store APIs, providing both efficiency and predictability
+> - Abstracts away filesystem and operating system specific quirks, ensuring portability
+> - Allows for functionality not native to filesystems, such as operation preconditions and atomic multipart uploads
+
+Where possible, implementations should use the underlying `obstore` APIs
+directly. Only where this is not possible should users fall back to this fsspec
+integration.
+"""
+
+from __future__ import annotations
+
+import asyncio
+from collections import defaultdict
+from typing import Any, Coroutine, Dict, List, Tuple
+
+import fsspec.asyn
+import fsspec.spec
+
+import obstore as obs
+
+
+class AsyncFsspecStore(fsspec.asyn.AsyncFileSystem):
+    def __init__(
+        self,
+        store,
+        *args,
+        asynchronous=False,
+        loop=None,
+        batch_size=None,
+        **kwargs,
+    ):
+        self.store = store
+        super().__init__(
+            *args, asynchronous=asynchronous, loop=loop, batch_size=batch_size, **kwargs
+        )
+
+    async def _rm_file(self, path, **kwargs):
+        return await obs.delete_async(self.store, path)
+
+    async def _cp_file(self, path1, path2, **kwargs):
+        return await obs.copy_async(self.store, path1, path2)
+
+    async def _pipe_file(self, path, value, **kwargs):
+        return await obs.put_async(self.store, path, value)
+
+    async def _cat_file(self, path, start=None, end=None, **kwargs):
+        if start is None and end is None:
+            resp = await obs.get_async(self.store, path)
+            return await resp.bytes_async()
+
+        if start is not None and end is not None:
+            return await obs.get_range_async(
+                self.store, path, offset=start, length=end - start
+            )
+
+        raise NotImplementedError("todo: handle open-ended ranges")
```

Comment:
This was re-implemented in https://github.com/developmentseed/obstore/pull/67, with clearer semantics for offset and suffix requests. Would you be able to map the fsspec range semantics onto this?

Reply:
I just pushed what I have at the moment, so can do whatever you like. 

Note that in the current state of the code, ranges_one and ranges_two both pass separately, but not when run in the same session ("Generic S3 error" on the upload), and I don't know why.

Reply:
Right now in the code on main there's no clearing of the moto state between test fixtures. I had run into some issue with that. So that could be a place to start.